sdks/python/apache_beam/transforms/external.py

```diff
@@ -42,6 +44,63 @@
   grpc = None
 # pylint: enable=wrong-import-order, wrong-import-position, ungrouped-imports
 
+DEFAULT_EXPANSION_SERVICE = 'localhost:8097'
+FILTERED_CODERS = {
+  'beam:coder:length_prefix:v1',
+}
+
+
+def iter_urns(coder, context=None):
+  yield coder.to_runner_api_parameter(context)[0]
+  for child in coder._get_component_coders():
+    for urn in iter_urns(child, context):
+      yield urn
+
+
+class External(ptransform.PTransform):
```

Comment:
Naming nit: it seems odd to have a transform named External and another named ExternalTransform. Maybe call this ExternalWrapperTransform? Or maybe, given that this is still all relatively experimental, this should all be folded into the same class?

Reply:
Also agree here.  I struggled with the same questions.  

Another option is to rename the current `ExternalTransform` to `_ExternalTransform`, since that's part of the API we'll be hiding away from the end user, and I'd rather keep the short-and-sweet name for the public.

I'm also fine with changing the existing `ExternalTransform` to become the base class, but it'll require a lot more adjustments to the tests.

Reply:
I also think `External` would be nice. IMHO Consolidating the two classes would only be confusing. Note that in Java we use `External` for `ExternalTransform`. A bit of a naming mess.

What about `ExternalTransformBase`? 

Or could we use a builder pattern instead?

Reply:
I'm going to come back to the naming question at the end.  For now I want to focus on the design.

Reply:
+1 for simplifying. I prefer folding everything to a single class named 'External'. I don't see the need to have two base classes if that is going to be overridden anyways. Also this will fix the naming discrepancy between Java and Python.

Reply:
I can do this, but just to be clear, `External` would become an abstract base class in that case, in order to port over the changes I've made to simplify the interface.

So this:

```python
class ExternalTransform(ptransform.PTransform):
  _namespace_counter = 0
  _namespace = threading.local()

  _EXPANDED_TRANSFORM_UNIQUE_NAME = 'root'
  _IMPULSE_PREFIX = 'impulse'

  def __init__(self, urn, payload, endpoint):
    if grpc is None and isinstance(endpoint, str):
      raise NotImplementedError('Grpc required for external transforms.')
    self._urn = urn
    self._payload = payload
    self._endpoint = endpoint
    self._namespace = self._fresh_namespace()

  [...]
```

Would become roughly:

```python
class External(ptransform.PTransform):
  _urn = None
  _namespace_counter = 0
  _namespace = threading.local()

  _EXPANDED_TRANSFORM_UNIQUE_NAME = 'root'
  _IMPULSE_PREFIX = 'impulse'

  def __init__(self, endpoint):
    if grpc is None and isinstance(endpoint, str):
      raise NotImplementedError('Grpc required for external transforms.')
    if self._urn is None:
      raise NotImplementedError('External sub classes must provide a _urn attribute.')
    self._endpoint = endpoint
    self._namespace = self._fresh_namespace()
    self._schema = self.get_schema()

  [...]
```

Such that we can write a subclass like:

```python
class ReadFromKafka(External):
  byte_array_deserializer = 'org.apache.kafka.common.serialization.' \
                            'ByteArrayDeserializer'

  _urn = 'beam:external:java:kafka:read:v1'

  def __init__(self, 
               consumer_config,
               topics,
               key_deserializer=byte_array_deserializer,
               value_deserializer=byte_array_deserializer,
               expansion_service=None):
    super(ReadFromKafka, self).__init__(expansion_service)
    self.consumer_config = list(consumer_config.items())
    self.topics = topics
    self.key_deserializer = key_deserializer
    self.value_deserializer = value_deserializer

  def get_schema(self):
    return {
    'consumer_config': typehints.Iterable[typehints.KV[unicode, unicode]],
    'topics': typehints.Iterable[unicode],
    'key_deserializer': unicode,
    'value_deserializer': unicode,
  }

```

This will change the tests a bit. 

Does that all sound good?

Reply:
For the sake of keeping separation of concerns and readability of the code, I would not advise merging the two. One class is responsible for **resolving an External transform via the expansion server**, the other class is responsible for **building a type of payload**. Merging the two will make it harder to add other ways to build payload for external transforms, e.g. add an explicit Proto message for certain transforms.

The abstract class would make this less worse, but the Beam pattern is to let a transform expand. I think we will benefit from the decoupling. If we want to fix the naming issues, we can just rename the classes like already suggested.

So just to recap:

Java: There is only the raw `External` which is equivalent to `ExternalTransform` in Python
Python: There is `External` (Payload building) and `ExternalTransform` (Resolution via Expansion Service)

How about renaming External to `ExternalWithSchema`? To consolidate with Java we could then also rename `ExternalTransform` to `External`.

Reply:
I think I agree with @mxm.  I'm still a bit new to beam, but I think combining them could add quite a lot of clutter to a class that's meant to be subclassed by users. 

As for names, if we care more about a nice python interface, rather than parity with Java, I might go with:

- rename raw `ExternalTransform` to `_External`: since most users will not directly interact with this
- rename payload building `External` to `ExternalTransform`: since this has the cleanest most self-explanatory name

If we care about parity with Java:

- rename raw `ExternalTransform` to `External`
- rename payload building `External` to `ExternalTransformInterface`

Reply:
I guess a part of my question was whether we need an inheritance hierarchy to do this or can payload building be a utility/delegate that is not necessarily a transform. Having a similarly named two transforms could be confusing to users. How about adding an ExternalTransformPayloadBuilder that can be sub-classed by users (with a default implementation) that is used as a utility by the transform "External" or "ExternalTransform " that all external transforms have to extend ?

In general we try to keep SDKs consistent but this doesn't have to come at the cost of making Python API not nice.

Reply:
> I guess a part of my question was whether we need an inheritance hierarchy to do this or can payload building be a utility/delegate that is not necessarily a transform. 

I'll break my choice down into a set of user stories / requirements.

**User story 1: As a developer writing a pipeline, using an external transform in a pipeline should _feel_ no different than using a native transform.**

For example, currently we're using `ReadFromPubSub`, but it's not supported on Flink.   Upgrading the following code to an external transform should be as easy as replacing the import:

```python
from apache_beam.io.gcp.pubsub import ReadFromPubSub
pipe = beam.Pipeline(options=pipeline_options)
pipe  | 'PubSubInflow' >> ReadFromPubSub(
    topic='projects/foo/topics/output', 
    with_attributes=True)
```

**User story 2: As a developer writing an external transform in python, it should be as easy as providing a schema and a URN.**   

I've given a few examples of this in this PR, using both type annotations and dataclasses.

---

I think inheritance is the best approach to meeting both of these requirements.  If we can agree on these two requirements, then we just need to determine whether it is also a requirement to have a lower-level concrete class like `ExternalTransform`.   If it is, then we need 2 classes, otherwise, we can use 1.

Unless anyone objects, I think I'll make an adjustment to the PR to combine the functionality of `External` into `ExternalTransform` so that we can see what the design would look like with a single abstract class.

Reply:
User story 1:...
Completely agree.
Minor point: Python ReadFromPubSub is not supported in Flink since it's a native transform (fully implemented in Dataflow runner). But yeah, using Java custom PubSub source from Python as a cross-language transform will give us something that works for all runners.

User story 2:...
Also agree. But we have to figure out what is the write API to provide config, schema, URN.
Agreeing with previous comment from @mxm, we have two separate features related to cross-language transforms that we would like to keep in separate classes.
(1) An composite transform that is expandable through the expansion service.
(2) An abstraction for building the Payload for (1).

Looking at these two (1) is surely a Beam transform, but looking at (2) it sounds like something else. So what I'm proposing is.

```python
class External(ptransform.PTransform):
  // We can call this External or ExternalTransform. This is the class that performs expansion.

  // URN is expected to be set by subclasses.
  urn = None
  
 
  def get_payload_builder():
     raise NotImplementedError

class PayloadBuilder:
  def get_conded_config():
    pass

class SchemaBasedPayloadBuilder(PayloadBuilder): 
  def get_schema():
    pass

```
In other words, we can maintain an inheritance hierarchy for payload building. But this doesn't have to be an inheritance hierarchy of Beam transforms.

Reply:
> Minor point: Python ReadFromPubSub is not supported in Flink since it's a native transform (fully implemented in Dataflow runner). But yeah, using Java custom PubSub source from Python as a cross-language transform will give us something that works for all runners.

full disclosure: I'm currently working on making `PubSubIO` externalizable, which is what lead me down this path in the first place.

> In other words, we can maintain an inheritance hierarchy for payload building. But this doesn't have to be an inheritance hierarchy of Beam transforms.

yes, but unless I'm missing something, we can't accomplish user story 1 with this approach. 
 Consider this hypothetical example:

```python
from apache_beam.io.external.gcp.pubsub import ReadFromPubSubPayloadBuilder
pipe = beam.Pipeline(options=pipeline_options)
# ERROR: payload builder is not a transform!
pipe  | 'PubSubInflow' >> ReadFromPubSubPayloadBuilder(
    topic='projects/foo/topics/output', 
    with_attributes=True)
```

Reply:
Actually, what I meant was something like.

```python
class ReadFromPubSubPayloadBuilder(SchemaBasedPayloadBuilder):
  def __init__(self, topic):
    self._topic = topic

  def get_schema(self):
    // Implementation

# We can call this class just ReadFromPubSub in case we want to make cross-language version the default.
class ReadFromPubSubExternal(ExternalTransform):
 
  URN = 'urn:external:pubsub'

  def __init__ (self, topic):
    self._payload_builder = ReadFromPubSubPayloadBuilder(topic)

  def get_payload_builder(self):
    return self._payload_builder:

  def expand(self, input):
    # optionally build a composite transform by combining ExternalTransform.expand() with other transforms.
```

#### User pipeline
```python
from apache_beam.io.external.gcp.pubsub import ReadFromPubSubExternal
pipe = beam.Pipeline(options=pipeline_options)
pipe  | 'PubSubInflow' >> ReadFromPubSubExternal(
    topic='projects/foo/topics/output', 
    with_attributes=True)
```

Reply:
@chamikaramj  Ah, I get it now.    

In your example, probably the payload builder will need to be instantiated with `self` (the `ExternalTransform` instance) to get the complete context, but yeah, should be doable.  I'll play around with that idea to see if I can some up with something that meets my UX requirements for User Story 2.

Reply:
Yeah, prob. `SchemaBasedPayloadBuilder` should be instantiated with self (transform). I'd say leave `PayloadBuilder` mostly unimplemented for someone who wishes to build the payload in a completely different way.

Reply:
I pushed a change that implements this new design.  I think it's a big improvement. 

One last design question:  should `ExternalTransform.__init__` accept a payload builder, or a payload?  It would simplify the interface for writers of external transforms a little bit, but removing the need to call `.build()`.

Here's a brief overview of the main payload builders that are available (in order of increasing brevity):

## NamedTuple

```python

ReadFromKafkaSchema = typing.NamedTuple(
  'ReadFromKafkaSchema',
  [
    ('consumer_config', typing.List[typing.Tuple[unicode, unicode]]),
    ('topics', typing.List[unicode]),
    ('key_deserializer', unicode),
    ('value_deserializer', unicode),
  ]
)


class ReadFromKafka(ExternalTransform):

  byte_array_deserializer = 'org.apache.kafka.common.serialization.' \
                            'ByteArrayDeserializer'

  URN = 'beam:external:java:kafka:read:v1'

  def __init__(self, consumer_config,
               topics,
               key_deserializer=byte_array_deserializer,
               value_deserializer=byte_array_deserializer,
               expansion_service=None):
    super(ReadFromKafka, self).__init__(
      self.URN,
      NamedTupleBasedPayloadBuilder(
        ReadFromKafkaSchema(
          consumer_config=list(consumer_config.items()),
          topics=topics,
          key_deserializer=key_deserializer,
          value_deserializer=value_deserializer,
        )
      ).build(),
      expansion_service
    )
```

## Annotations

```python
class ReadFromKafka(ExternalTransform):
  byte_array_deserializer = 'org.apache.kafka.common.serialization.' \
                            'ByteArrayDeserializer'

  URN = 'beam:external:java:kafka:read:v1'

  def __init__(self, consumer_config: typing.List[typing.Tuple[unicode, unicode]],
               topics: typing.List[unicode],
               key_deserializer: unicode = byte_array_deserializer,
               value_deserializer: unicode = byte_array_deserializer,
               expansion_service=None):

    super(ReadFromKafka, self).__init__(
      self.URN,
      AnnotationBasedPayloadBuilder(
        self,
        consumer_config=consumer_config,
        topics=topics,
        key_deserializer=key_deserializer,
        value_deserializer=value_deserializer,
      ).build(),
      expansion_service
    )
```

## Dataclasses

We should be able to simplify this even further with a class decorator or base class to implement `__post_init__`.

```python
@dataclass
class ReadFromKafka(ExternalTransform):
  byte_array_deserializer = 'org.apache.kafka.common.serialization.' \
                            'ByteArrayDeserializer'

  URN = 'beam:external:java:kafka:read:v1'

  consumer_config: typing.List[typing.Tuple[unicode, unicode]]
  topics: typing.List[unicode]
  key_deserializer: unicode = byte_array_deserializer
  value_deserializer: unicode = byte_array_deserializer
  expansion_service: InitVar[typing.Optional[str]] = None

  def __post_init__(self, expansion_service):
    super(ReadFromKafka, self).__init__(
      self.URN,
      DataclassBasedPayloadBuilder(self).build(),
      expansion_service
    )
```

Reply:
I updated `ExternalTransform` to accept either a payload or a builder.